dynect/src/test/java/denominator/dynect/DynECTFunctionsTest.java

```diff
@@ -0,0 +1,44 @@
+package denominator.dynect;
+
+import com.google.gson.*;
+import denominator.common.Util;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Map;
+
+import static org.testng.Assert.assertEquals;
+
+public class DynECTFunctionsTest{
```

Comment:
I understand that you are trying to not repeat yourself, but this class is very indirect. Unit tests should favor readability and straight-forwardness above line count. When we get meta, we lose the purpose of the file and force users to scroll through multiple files and boilerplate concepts to understand what's going on. This goes against the direction of the work I've been on, which has been to simplify how our tests behave.

Try the following, which is very straight-forward and requires no scrolling.

```java
  @Test
  public void aaaaRecord() throws Exception {
    assertThat(toRData("{\n"
        + "    \"zone\": \"denominator.io\",\n"
        + "    \"ttl\": 3600,\n"
        + "    \"fqdn\": \"www.denominator.io\",\n"
        + "    \"record_type\": \"AAAA\",\n"
        + "    \"rdata\": {\n"
        + "      \"address\": \"2001:db8::3\"\n"
        + "    },\n"
        + "    \"record_id\": 2\n"
        + "  }"))
        .isInstanceOf(AAAAData.class)
        .containsEntry("address", "2001:db8::3")
    // or
        .isEqualTo(AAAAData.create("2001:db8::3"));
  }

  private static Map<String, Object> toRData(String inputJson) {
    return ToRecord.toRData(new JsonParser().parse(inputJson).getAsJsonObject());
  }
```

Or the more meta, but still much less difficult to understand parameterized version

```java

  @DataProvider(name = "rdata") public static Object[][] rdata() {
    return new Object[][] {
        {
            "{\n"
          + "    \"zone\": \"denominator.io\",\n"
          + "    \"ttl\": 3600,\n"
          + "    \"fqdn\": \"www.denominator.io\",\n"
          + "    \"record_type\": \"AAAA\",\n"
          + "    \"rdata\": {\n"
          + "      \"address\": \"2001:db8::3\"\n"
          + "    },\n"
          + "    \"record_id\": 2\n"
          + "  }",
            AAAAData.create("2001:db8::3")
        } // repeat for each rdata to test
    };
  }

  @Test(dataProvider = "rdata")
  public void toRData(String inputJson, Map<String, Object> expectedResult) {
    assertThat(ToRecord.toRData(new JsonParser().parse(inputJson).getAsJsonObject()))
        .isInstanceOf(expectedResult.getClass())
        .isEqualTo(expectedResult);
  }
```

Hope this helps!